pkg/controllers/cluster/kubeconfig_controller.go

```diff
@@ -0,0 +1,257 @@
+// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Greenhouse contributors
+// SPDX-License-Identifier: Apache-2.0
+
+package cluster
+
+import (
+	"context"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/clientcmd"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+
+	greenhouseapis "github.com/cloudoperators/greenhouse/pkg/apis"
+	"github.com/cloudoperators/greenhouse/pkg/apis/greenhouse/v1alpha1"
+	"github.com/cloudoperators/greenhouse/pkg/clientutil"
+)
+
+type KubeconfigReconciler struct {
+	client.Client
+}
+
+//+kubebuilder:rbac:groups=greenhouse.sap,resources=clusters,verbs=get;list;watch
+//+kubebuilder:rbac:groups=greenhouse.sap,resources=organizations,verbs=get;list;watch
+//+kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch
+// +kubebuilder:rbac:groups=greenhouse.sap,resources=cluster-kubeconfigs,verbs=get;list;watch;create;update;patch
+
+// SetupWithManager sets up the controller with the Manager
+func (r *KubeconfigReconciler) SetupWithManager(name string, mgr ctrl.Manager) error {
+	r.Client = mgr.GetClient()
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		Watches(&v1alpha1.Cluster{}, handler.EnqueueRequestsFromMapFunc(sameNameResource)).
+		Watches(&corev1.Secret{}, handler.EnqueueRequestsFromMapFunc(clusterSecretToCluster)).
+		Watches(&v1alpha1.Organization{}, handler.EnqueueRequestsFromMapFunc(r.organizationToClusters)).
+		Watches(&corev1.Secret{}, handler.EnqueueRequestsFromMapFunc(r.organizationSecretToClusters)).
+		For(&v1alpha1.ClusterKubeconfig{}).
+		Complete(r)
+}
+
+func (r *KubeconfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	l := log.FromContext(ctx).WithValues("cluster", req.Name, "namespace", req.Namespace)
+
+	var cluster v1alpha1.Cluster
+	if err := r.Get(ctx, req.NamespacedName, &cluster); err != nil {
+		if client.IgnoreNotFound(err) != nil {
+			return ctrl.Result{}, err
+		}
+		l.Info("skip reconcile, cluster is not found")
+		return ctrl.Result{}, nil
+	}
+
+	if cluster.GetDeletionTimestamp() != nil {
+		l.Info("skip reconcile, cluster is being deleted")
+		return ctrl.Result{}, nil
+	}
```

Comment:
Additionally to Cluster found and Cluster not in deletion we should check Cluster `Ready` StatusCondition.
If ClusterStatus is not ready, we should persist into a StatusCondition of the Kubeconfig (e.g. `ClusterReady`) or skip Reconciliation at all. WDYT @IvoGoman

Reply:
Agreed on the check of Cluster `Ready` but maybe its sufficient to have one status condition for that reflects the reconciliation success with different reasons for the error conditions. This would make it easier to grasp in the UI. (especially once we get mouse overs)

Reply:
Cluster ready status is being checked by 16b7fb7af545499f90a2f168752ee2ceb335a60a and reconciliation is skipped when `Cluster` is not `Ready` (not to change kubeconfig data while the cluster is in action)